pkg/proxy/handler.go

```diff
@@ -83,13 +87,36 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 
 	log.Println("Waiting for IPs", id)
-	host := h.Endpoints.GetHost(r.Context(), deploy, "http")
+	host, err := h.Endpoints.AwaitHostAddress(r.Context(), deploy, "http")
+	if err != nil {
+		log.Printf("timeout finding the host address %v", err)
+		w.WriteHeader(http.StatusRequestTimeout)
+		w.Write([]byte(fmt.Sprintf("Request timed out for model: %v", modelName)))
+		return
+	}
 	log.Printf("Got host: %v, id: %v\n", host, id)
 
 	// TODO: Avoid creating new reverse proxies for each request.
 	// TODO: Consider implementing a round robin scheme.
 	log.Printf("Proxying request to host %v: %v\n", host, id)
-	newReverseProxy(host).ServeHTTP(w, proxyRequest)
+	middleware := withCancelDeadTargets(h.Endpoints, deploy, host)
+	middleware(newReverseProxy(host)).ServeHTTP(w, proxyRequest)
+}
+
+func withCancelDeadTargets(endpoints *endpoints.Manager, deploy string, host string) func(other http.Handler) http.HandlerFunc {
+	return func(other http.Handler) http.HandlerFunc {
+		return func(w http.ResponseWriter, r *http.Request) {
+			newCtx, done, err := endpoints.RegisterInFlight(r.Context(), deploy, host)
+			if err != nil {
+				log.Printf("error registering in-flight request: %v", err)
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			defer done()
+
+			other.ServeHTTP(w, r.Clone(newCtx))
```

Comment:
I don't understand where the retry for 503 is happening in the middleware. Could you explain how the retry mechanism works?

I was looking for some part of the code that would catch an error response (e.g. all 5xx errors) and then send a new request to a new endpoint.

Reply:
The retry is not implemented yet. This is the first part where an inflight proxy request is canceled. I will push some changes today to add retry middleware